### ISSUE TITLE
Accept Kotlin code?

### ISSUE THREAD

**Issue by fab1an** [glazedlists/glazedlists#654]
Is anyone opposed to allowing kotlin code in the glazedlist-source?

For who hasn‘t looked at it yet: It‘s 100% compatible with Java, also rather readable if you know Java. The files can live in the same source folder, Java can call Kotlin classes and the otherway round. 

The only drawback I can think of is an additional dependency (the stdlib).

**Comment by robeden** [comment]
I wouldn't want to require the stdlib as a dependency, personally.

The main reason to push for it, I would think, would be to pull closures and such into Java 7, but since new GL work will push to Java 8 it seems unnecessary. Why would we want to do this other from a technical point of view? (I understand it may be what some people are mostly working in these days and it has a high coolness factor at the moment... so I get those reasons. 😉 )

**Comment by fab1an** [comment]
I understand ;).
In general I say the less dependencies the better, especially if it's on badly maintained libraries. 

Allow me to do some advertising nevertheless: I use Kotlin since three years now so I know it in depth. The features that help development the most aren't even lambdas in my opinion (which are Java6 compatible btw.!), but these:

*Readability*
- Much less boilerplate. Through type-inference there are seldomly type-annotation necessary, so `val str = "test"` instead of `final String str = "test"
- One line `class MyClass(private val x: Int)` instead of `class MyClass { constructor(x: Int) {}... this.x=x}` etc.
- If you swap `class` for `object` the class becomes a static singleton
- If you use `data class` instead of `class` you get `equalsTo` and `hashCode` implemented via all of the values in the constructor.
- Extension functions, so I can write for example "myEventList.newDerivedBasicEventList()" to return a list with publisher and locks shared.

*Code safety* 
- Having null in the type-system so compile-time safety against NullPointerExceptions
- `var`/`val` variables instead of the `final`keyword everywhere
- Many useful stdlib operations, like `List<T>.single(): T`, which returns T if it's the single element in the list and throws otherwise.
- Range operations (0..list.lastIndex).
- No automatic casting. Conversion between numbers requires explicit conversion via ".toLong()." et al. 

I can come up with more, I think ;).

**Comment by robeden** [comment]
Yep. I like Kotlin a lot. But, just to be clear: none of that is a technical requirement to this project.

I'll defer to the group decision.

**Comment by hbrands** [comment]
I'm also interested in Kotlin but I'm still a newbie so to say. So, at this point I would not feel comfortable to adopt Kotlin in Glazed Lists (yet). And from a (mainstream) user perspective I don't see an urgent need for it either.
But we should rethink supporting/using Kotlin for a major revision of Glazed Lists.

**Comment by fab1an** [comment]
Thanks for your comments. In principle, with some work done, I think it should be possible to use it without stdlib. 
The needed extensions functions line the aforementioned „.single()“ or „.forEach“ could be copied even, they are mostly one line functions that delegate to Iterators and get inlined anyway. 

There are some wrappers around java types, but I think it should be possible to work without them. 

What i would favor as long term solution is a very minimal core of GL: BasicEventList, FilterList, SortedList, MatcherEditor without locking/threading or utility functions in kotlin so I can compile to JavaScript and native too. Maybe GroupingList and CompositeList too.

I work on various platforms and sadly especially the JS-world hasn‘t yet found out how easy and performant list-transformations could be.

If you know the ListEvent interface and see what, for example, react-native developers keeping hacking together and throwing away again to get somewhat peformant lists... 😮

**Comment by fab1an** [comment]
There is another possible way, as soon as there’s actual submitted code in kotlin: we could put it into it’s own module, gradle separates it from the rest. 

It would be excluded from the normal jar, but published as a separate artifact. unittests run as usual. If someone wants to use the code they can add the additional jar and over time it could be translated piecewise to java and moved to the core. 
Just theory of course.

**Comment by fab1an** [comment]
I have some half-baked code I’d like to rewrite and add when I come around to: i call it EventGrid. 

The base problem was that i needed a fast grid view of an eventlist with millions of items.  

Since there are functional O(1) renderers like react nowadays, i implemented it like this:

The EventGrld is a ListEventListener and has a function to determine height/width of a listitem, which is usually fast. 

At any time you can set the viewport size in pixels and the current scrollX/Y offset on the Grid. 

There is a property/getter which will provide you with a window of the currently visible items, each item <T>, it‘s height/width and it‘s relative x/y position in the viewport. Also the total height of the window is provided for correct scrollbars. 

You can then pipe that into react or anything that can draw pixels to update the screen with the window. 

Benefits
• table view is just a special case where all items have the width of the window, you can also have varying column width or different height. 
• The grid never knows about the actual content, just it‘s bounds, so you can map the provided window and display arbitray information in the bounds. 
• arbitrary sizes are possible for separators etc. 
• in sorted lists it also support sticky section headers that stay in the top of the viewport until their section is gone, when they scroll up to make room for the next header

I think it‘s a great abstraction, since porting it to any UI just requires a fast update/render function that can 
• set width/height and scrollbar position on a window, in react i do that by painting a large white div element. 
• put/paint items in there using viewport relative pixel coordinates. 

I think it would also be possible to implement native widgets that way and simplify EventTableViewer or improve it‘s performance.

**Comment by fab1an** [comment]
That being said, since the code isn‘t reusable in it‘s current form anyway, i had an idea to improve on that:

Since most realworld performance is lost on sorting the table, I got the idea of building incremental sorting into the Grid. Tapping a column sorts, but the grid only actually needs to find the top n elements that are visible, which can be done in a single pass. Scrolling down would sort more of the results. Since this is all about viewing and the UI, one could also provide a frame-time-limit to the grid like 16ms which it can use to presort, before aborting.

**Comment by robeden** [comment]
It sounds like a good project to me. I've implemented basic rendering bits (not the incremental sorting) several times back in my Swing days.

I would personally like to see extensions that are distributed as separate maven artifacts and this might be a great first thing to dip our toes in with that approach.
In the separate artifacts case, I think Kotlin is a perfectly fine thing to use. My only "concern" is currently with "core" code.
